include/Gaffer/Private/SlotBase.h

```diff
@@ -0,0 +1,163 @@
+//   the signal and vice-versa.
+// - Robustness to interesting situations such as self-disconnecting slots and
+//   slots that cause the signal itself to be deleted.
+// - Minimal storage requirements among the various Signals components.
+//
+// To meet these requirements, SlotBase objects are reference counted and form a
+// doubly linked list with a few specific features :
+//
+// - The next slot is referenced by `intrusive_ptr`. This allows Signal to
+//   maintain the lifetime of the slots purely by holding a single pointer to
+//   the first slot, and also allows Connection to share ownership via it's own
+//   `intrusive_ptr`.
+// - The `previous` field doesn't point to the previous _slot_, but instead to
+//   its `SlotBase::next` field. This allows the first slot to point back
+//   directly to `Signal::m_firstSlot`, meaning we can remove a slot from the
+//   list without needing access to a Signal object.
+// - Disconnected slots keep a valid `next` pointer, so that SlotCallIterator
+//   can continue iteration even if a slot disconnects itself when called.
+//
+struct SlotBase : private boost::noncopyable
+{
+
+	using Ptr = boost::intrusive_ptr<SlotBase>;
+
+	// Inserts slot after `previous`.
+	SlotBase( Ptr &prev )
+		:	previous( &prev ), next( prev ), m_referenceCount( 0 ), blocked( false ), calling( false )
+	{
+		*previous = this;
+		if( next )
+		{
+			next->previous = &next;
+		}
+	}
+
+	virtual ~SlotBase()
+	{
+	}
+
+	// Removes slot from list. Virtual so that Signal::Slot
+	// can also destroy its slot functor.
+	virtual void disconnect()
+	{
+		if( !previous )
+		{
+			// Already disconnected
+			return;
+		}
+
+		if( next )
+		{
+			next->previous = previous;
+		}
+		Ptr &previousRef = *previous;
+		assert( previousRef == this );
+		previous = nullptr;
+		// This assignment last, since it could drop
+		// our reference count to 0 and destroy us.
+		previousRef = next;
+		// Note : We have deliberately kept our pointer to `next`, so that we
+		// don't invalidate the SlotCallIterator when a slot disconnects itself
+		// from within a call. We're not reachable from the head of the list,
+		// but the end of the list is still reachable from us.
+	}
+
+	// Non-null when connected to signal (reachable from `Signal::m_firstSlot`).
+	Ptr *previous;
+	Ptr next;
+
+	// Reference count. We don't derive from IECore::RefCounted or
+	// `boost::intrusive_ref_counter` for several reasons :
+	//
+	// - We don't need an atomic reference count, because signals are not
+	//   intended to be threadsafe. Using a non-atomic count gives us
+	//   significantly improved performance.
+	// - We don't need a 64 bit count, and using a smaller type stored _after_
+	//   our pointer fields lets us minimise `sizeof( SlotBase )`.
+	uint32_t m_referenceCount;
+	bool blocked;
+	// True when currently in `Slot::operator()`. Used to defer destruction
+	// of self-disconnecting functions.
+	bool calling;
```

Comment:
As I understand it, struct padding means that `sizeof( SlotBase )` must be a multiple of 8 (sizeof largest member). That means the size isn't reduced even if `blocked/calling` are removed completely, because they are sitting inside the padding after the 4-byte reference count. I'm not super-hot on padding/alignment rules, but I have verified this experimentally.